internal/utils/time_interval.go

```diff
@@ -0,0 +1,130 @@
+package utils
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+)
+
+const (
+	// ErrEndBeforeStart is an the error message for when a TimeInterval's end time
+	// would be before its start.
+	ErrEndBeforeStart = "end time before start time"
+
+	errWindowTooLargeFmt = "random window equal to or larger than TimeInterval: window %v, interval %v"
+)
+
+// TimeInterval represents an interval of time in UTC. That is, regardless of
+// what timezone(s) are used for the beginning and end times, they will be
+// converted to UTC and methods will return them as such.
+type TimeInterval struct {
+	start time.Time
+	end   time.Time
+}
+
+// NewTimeInterval creates a new TimeInterval for a given start and end. If end
+// is a time.Time before start, then an error is returned.
+func NewTimeInterval(start, end time.Time) (*TimeInterval, error) {
+	if end.Before(start) {
+		return nil, fmt.Errorf(ErrEndBeforeStart)
+	}
+	return &TimeInterval{start.UTC(), end.UTC()}, nil
+}
+
+// Duration returns the time.Duration of the TimeInterval.
+func (ti *TimeInterval) Duration() time.Duration {
+	return ti.end.Sub(ti.start)
+}
+
+// Overlap detects whether the given TimeInterval overlaps with this
+// TimeInterval, assuming an inclusive start boundary and exclusive end
+// boundary.
+func (ti *TimeInterval) Overlap(other *TimeInterval) bool {
+	thisS := ti.Start()
```

Comment:
I feel `thisS` and `thatS` are not reader-friendly. Easily confused, hard to see which is which on a glance. Something like `firstS` and `secondS` would be much better IMO.

Reply:
Fair point, I'll change to `s1`/`s2` and `e1`/`e2` since it's shorter